src/bus.rs

```diff
@@ -1,54 +1,165 @@
-use crate::joypad::Joypad;
+use crate::traits::{Device, Memory};
 
 const RAM_MIRRORING_MASK: u16 = 0b0000_0111_1111_1111;
 const PPU_REGISTERS_MIRRORING_MASK: u16 = 0b0010_0000_0000_0111;
 
+enum RunMode {
+    Read,
+    Write,
+}
+
 pub struct Bus {
-    pub(crate) memory:  *mut [u8; 0xFFFF],
-    pub(crate) joypad_1: *mut Joypad,
-    pub(crate) joypad_2: *mut Joypad,
+    memory: *mut [u8; 0xFFFF],
+    devices: Vec<(std::ops::Range<usize>, *mut dyn Device)>,
 }
 
 impl Bus {
-    pub fn new(
-        memory: *mut [u8; 0xFFFF],
-        joypad_1: *mut Joypad,
-        joypad_2: *mut Joypad,
-    ) -> Self {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
         Self {
-            memory,
-            joypad_1,
-            joypad_2,
+            memory: std::ptr::null_mut(),
+            devices: Vec::new(),
         }
     }
 
     #[allow(clippy::missing_safety_doc)]
-    pub unsafe fn mem_read_u8(&mut self, addr: u16) -> u8 {
-        let addr = mirror_address(addr);
-        match addr {
-            0x4016 => (*self.joypad_1).write_mem(),
-            0x4017 => (*self.joypad_2).write_mem(),
-            _ => todo!(),
+    pub unsafe fn map(&mut self, memory: &mut [u8; 0xFFFF], devices: *const [*mut dyn Device]) {
+        self.memory = memory;
+        for device in (*devices).iter() {
+            let mapping = (**device).mapping_def();
+            (**device).map(&mut (*self.memory)[mapping.clone()]);
+            self.devices.push((mapping, *device));
+        }
+    }
+
+    fn run_device(&mut self, address: u16, run_mode: RunMode) {
+        for (range, device) in self.devices.iter() {
+            if range.contains(&usize::from(address)) {
+                unsafe {
+                    match run_mode {
+                        RunMode::Read => (**device).mem_read(),
+                        RunMode::Write => (**device).mem_write(),
+                    }
+                }
+                break;
+            }
         }
+    }
```

Comment:
i don't see the reason for it to be a single function with 2 modes instead of 2 separate functions

Reply:
To avoid duplication the code which check for the range and call the function. We could pass a lambda rather than a mode if you prefer